index.d.ts

```diff
@@ -0,0 +1,77 @@
+// Type definitions for express-session 1.16
+// Project: https://github.com/expressjs/session
+// Definitions by: Hiroki Horiuchi <https://github.com/horiuchi>
+//                 Jacob Bogers <https://github.com/jacobbogers>
+//                 Naoto Yokoyama <https://github.com/builtinnya>
+//                 Ryan Cannon <https://github.com/ry7n>
+//                 HoldYourWaffle <https://github.com/HoldYourWaffle>
+
+import express from 'express'
+import node from 'events'
+
+
+
+declare global {
+  namespace Express {
+    interface Request { //inject additional properties on express.Request
+      session?: Session;
+      sessionID?: string;
+    }
+  }
+
+  interface SessionData {
+    //seperate interface to allow application-specific declaration merging (see README)
+  }
+}
+
+
+
+interface SessionOptions {
+  secret: string | string[];
+
+  genid?(req: express.Request): string;
+  name?: string;
+  store?: Store;
+  cookie?: express.CookieOptions;
```

Comment:
Where can I look up how this is defined? Not sure if this is a 1:1 mapping with the options in this module or not.

Reply:
[Here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/56607c79bd4f6de9a8d77925e6e09aaf0cf6491a/types/express-serve-static-core/index.d.ts#L165)

Reply:
Is just the properties and their values documented in the README (https://github.com/expressjs/session#cookie) confusing? Like we can make them better, do you have any suggestions?

Reply:
Looking at the linked type and here, I think the following are the differences:

- The `expires` is just a `Date` object, not `boolean` as defined in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/56607c79bd4f6de9a8d77925e6e09aaf0cf6491a/types/express-serve-static-core/index.d.ts#L168
- As far as I'm aware there is no `encode` option accepted by this module but it is defined at https://github.com/DefinitelyTyped/DefinitelyTyped/blob/56607c79bd4f6de9a8d77925e6e09aaf0cf6491a/types/express-serve-static-core/index.d.ts#L173

Reply:
> Is just the properties and their values documented in the README (https://github.com/expressjs/session#cookie) confusing? Like we can make them better, do you have any suggestions?

Although the formatting could be better the README is quite clear from a user perspective. It's however hard to understand how the code's "type system" actually works (mainly because it's, well, JavaScript) which is pretty important if you're writing type definitions.

> Looking at the linked type and here, I think the following are the differences: ...

Updated. Just to be sure: does `secure` indeed accept a string "auto" or did you overlook that one?

Reply:
Yes, it accepts the string auto:

https://github.com/expressjs/session#cookiesecure

> The cookie.secure option can also be set to the special value 'auto' to have this setting automatically match the determined security of the connection.